### DISCUSSION TITLE
FillArc question/info/help

### DISCUSSION THREAD

**Discussion by Kotik82** [Bodmer/TFT_eSPI#1600] [Q&A - General]
Hello,

I recently installed the library and started experimenting with it in order to adapt it to my project.

My goal is to create a digital thermometer based on ESP8266, a GC9A01 tound ips display 240 x 240, the library itself is working fine, but i am just too stupid to figure out how to properly code what i got in my head.

TEMP DELAY is set to 800ms so in reality the tft is being updated once every 800ms, it all works great with 1 exception.... when the temp value increases the arc is filled up correctly, but when the temp value is decreasing the arc remains in its previous state. I guess it is cause the previous drawn pixel still remain on the screen.

Till now i got this code down:

```
bool updateTemperature() {
  if ((millis() - lastTempUpdate) > TEMP_READ_DELAY) {
    temperature = therm.getTempAverage(); // Get temperature reading
    lastTempUpdate = millis();
    tft.setTextColor(TFT_WHITE, TFT_BLACK);
    tft.setCursor(100, 60, 2);           // Start position
    tft.print(temperature);              // Shows value
    fillArc(120, 120, 225, temperature, 120, 120, 20, TFT_RED);
    return true;
  }
  return false;
}

void fillArc(int x, int y, int start_angle, int seg_count, int rx, int ry, int w, unsigned int colour)
{

  byte seg = 1; // Segments are 3 degrees wide = 120 segments for 360 degrees
  byte inc = 1; // Draw segments every 3 degrees, increase to 6 for segmented ring

  // Calculate first pair of coordinates for segment start
  float sx = cos((start_angle - 90) * DEG2RAD);
  float sy = sin((start_angle - 90) * DEG2RAD);
  uint16_t x0 = sx * (rx - w) + x;
  uint16_t y0 = sy * (ry - w) + y;
  uint16_t x1 = sx * rx + x;
  uint16_t y1 = sy * ry + y;

  // Draw colour blocks every inc degrees
  for (int i = start_angle; i < start_angle + seg * seg_count; i += inc) {

    // Calculate pair of coordinates for segment end
    float sx2 = cos((i + seg - 90) * DEG2RAD);
    float sy2 = sin((i + seg - 90) * DEG2RAD);
    int x2 = sx2 * (rx - w) + x;
    int y2 = sy2 * (ry - w) + y;
    int x3 = sx2 * rx + x;
    int y3 = sy2 * ry + y;

    tft.fillTriangle(x0, y0, x1, y1, x2, y2, colour);
    tft.fillTriangle(x1, y1, x2, y2, x3, y3, colour);

    // Copy segment end to sgement start for next segment
    x0 = x2;
    y0 = y2;
    x1 = x3;
    y1 = y3;
  }
}

void loop() {

updateTemperature();

}
```

So my question is how should i fix that issue?

Thanx

**Comment by Bodmer** [comment]
You need to store the last displayed value and redraw the segment in the range old value to new value. If the new value is less than the old value then blank the segment in between. If new value is bigger then colour the segment.

The ring meter uses a simple method of drawing the whole ring for every update but this does slow down updates.
